File: modules/expo-bluesky-video-compress/android/src/main/java/expo/modules/blueskyvideocompress/ExpoVideoCompressModule.kt
package expo.modules.blueskyvideocompress

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoVideoCompressModule : Module() {
  private var currentCompressor: VideoCompressor? = null

  override fun definition() =
    ModuleDefinition {
      Name("ExpoVideoCompress")

      Events("onProgress")

      AsyncFunction("probe") { uri: String ->
        val context = appContext.reactContext
          ?: throw Error("React context is null")
        return@AsyncFunction VideoProber.probe(context, uri)
      }

      AsyncFunction("compress") { uri: String, options: Map<String, Any?> ->
        val context = appContext.reactContext
          ?: throw Error("React context is null")
        val targetBitrate = (options["targetBitrate"] as? Number)?.toInt() ?: 3_000_000
        val maxSize = (options["maxSize"] as? Number)?.toInt() ?: 1920
        val jobId = (options["jobId"] as? Number)?.toInt() ?: 0

        val compressor = VideoCompressor(
          context = context,
          uri = uri,
          targetBitrate = targetBitrate,
          maxSize = maxSize,
          jobId = jobId,
          onProgress = { id, progress ->
            sendEvent("onProgress", mapOf(
              "id" to id,
              "progress" to progress
            ))
          }
        )

        currentCompressor = compressor

        try {
          val result = compressor.compress()
          currentCompressor = null
          return@AsyncFunction result
        } catch (e: Exception) {
          currentCompressor = null
          throw e
        }
      }

      Function("cancel") {
        currentCompressor?.cancel()
        currentCompressor = null
      }
    }
}

File: modules/expo-bluesky-video-compress/ios/ExpoVideoCompressModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoVideoCompressModule: Module {
  private var currentCompressor: VideoCompressor?

  public func definition() -> ModuleDefinition {
    Name("ExpoVideoCompress")

    Events("onProgress")

    AsyncFunction("probe") { (uri: String) -> [String: Any] in
      let url = URL(string: uri) ?? URL(fileURLWithPath: uri)
      return try await VideoProber.probe(url: url)
    }

    AsyncFunction("compress") { (uri: String, options: [String: Any]) -> [String: Any] in
      let url = URL(string: uri) ?? URL(fileURLWithPath: uri)
      let targetBitrate = options["targetBitrate"] as? Int ?? 3_000_000
      let maxSize = options["maxSize"] as? Int ?? 1920
      let jobId = options["jobId"] as? Int ?? 0

      let compressor = VideoCompressor(
        url: url,
        targetBitrate: targetBitrate,
        maxSize: maxSize,
        jobId: jobId,
        onProgress: { [weak self] id, progress in
          self?.sendEvent("onProgress", [
            "id": id,
            "progress": progress
          ])
        }
      )

      self.currentCompressor = compressor

      do {
        let result = try await compressor.compress()
        self.currentCompressor = nil
        return result
      } catch {
        self.currentCompressor = nil
        throw error
      }
    }

    Function("cancel") {
      self.currentCompressor?.cancel()
      self.currentCompressor = nil
    }
  }
}
